contrib/golang/filters/http/source/golang_filter.cc

```diff
@@ -1065,6 +1065,48 @@ void Filter::setDynamicMetadataInternal(ProcessorState& state, std::string filte
   state.streamInfo().setDynamicMetadata(filter_name, value);
 }
 
+CAPIStatus Filter::setStringFilterState(absl::string_view key, absl::string_view value,
+                                        int state_type, int life_span, int stream_sharing) {
+  // lock until this function return since it may running in a Go thread.
+  Thread::LockGuard lock(mutex_);
+  if (has_destroyed_) {
+    ENVOY_LOG(debug, "golang filter has been destroyed");
+    return CAPIStatus::CAPIFilterIsDestroy;
+  }
+
+  auto& state = getProcessorState();
+  if (!state.isProcessingInGo()) {
+    ENVOY_LOG(debug, "golang filter is not processing Go");
+    return CAPIStatus::CAPINotInGo;
+  }
+
+  if (state.isThreadSafe()) {
+    state.streamInfo().filterState()->setData(
+        key, std::make_shared<GoStringFilterState>(value),
+        static_cast<StreamInfo::FilterState::StateType>(state_type),
+        static_cast<StreamInfo::FilterState::LifeSpan>(life_span),
+        static_cast<StreamInfo::StreamSharingMayImpactPooling>(stream_sharing));
+  } else {
+    auto key_str = std::string(key);
+    auto value_str = std::string(value);
+    auto weak_ptr = weak_from_this();
+    state.getDispatcher().post(
+        [this, &state, weak_ptr, key_str, value_str, state_type, life_span, stream_sharing] {
+          if (!weak_ptr.expired() && !hasDestroyed()) {
+            Thread::LockGuard lock(mutex_);
+            state.streamInfo().filterState()->setData(
+                key_str, std::make_shared<GoStringFilterState>(value_str),
```

Comment:
2 memory copy, from string_view to std::string.

Reply:
yeah, but copyGostring for value may also need two copy when it's in the Go thread? one is copyGostring and the other is `GoStringFilterState`? and we could save a copy in C++ thread  using referGoString for value. 🤔

Reply:
AFAIK, std::string is refer count based, so that we only need one memory copy in C++ thread, if we use std::string all the time.
that means, the memory copy times depends on the string_view to std::string times.

Reply:
> AFAIK, std::string is refer count based

nope, every std::string object own itself's memory and will not shared it with anyone.

Reply:
A possible optimization is use move semantic when we catch the key/value in the closure.

Reply:
Oh, sorry for the big mistake, it's forbiden in c++11, and thanks for the correction.

another related question, both of these two styles need memory copy, for `value_(value)`, right? thanks again.
```
GoStringFilterState(absl::string_view value) : value_(value) {}
GoStringFilterState(std::string value) : value_(value) {}
```

Reply:
yeah, i think so. cc @wbpcode

Reply:
> Oh, sorry for the big mistake, it's forbiden in c++11, and thanks for the correction.
> 
> another related question, both of these two styles need memory copy, for `value_(value)`, right? thanks again.
> 
> ```
> GoStringFilterState(absl::string_view value) : value_(value) {}
> GoStringFilterState(std::string value) : value_(value) {}
> ```

I think the second style may require once additional copy (depend on the optimization of compiler).
```
GoStringFilterState(std::string value) : value_(std::move(value)) {} # May be better
```

Reply:
so, in summary, the realization now is better? since we can use `referGoString` for `value` and copy until `absl::string_view` to `std::string` in `GoStringFilterState`?

Reply:
or, maybe we could make_shared<GoStringFilterState> before `dispatcher.post`? so that there is only one copy for value string.

Reply:
oh, missing GoStringFilterState in the previous comment, I mean `make_shared< GoStringFilterState>`, does this useful?

Reply:
> oh, missing GoStringFilterState in the previous comment, I mean `make_shared< GoStringFilterState>`, does this useful?

yeah, i think it is useful.